ci/python-publish.yml

```
@@ -12,28 +12,42 @@ on:
   release:
     types: [published]
 
-permissions:
-  contents: read
-
 jobs:
-  deploy:
+  release-build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
 
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.x"
+
+      - name: build release distributions
+        run: |
+          # NOTE: put your own distribution build steps here.
+          python -m pip install build
+          python -m build
+
+      - name: upload distributions
+        uses: actions/upload-artifact@v4
+        with:
+          name: release-dists
+          path: dist/
+
+  pypi-publish:
     runs-on: ubuntu-latest
+    needs:
+      - release-build
+    permissions:
+      id-token: write
 
     steps:
-    - uses: actions/checkout@v3
-    - name: Set up Python
-      uses: actions/setup-python@v3
-      with:
-        python-version: '3.x'
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install build
-    - name: Build package
-      run: python -m build
-    - name: Publish package
-      uses: pypa/gh-action-pypi-publish@27b31702a0e7fc50959f5ad993c78deac1bdfc29
-      with:
-        user: __token__
-        password: ${{ secrets.PYPI_API_TOKEN }}
+      - name: Retrieve release distributions
+        uses: actions/download-artifact@v4
+        with:
+          name: release-dists
+          path: dist/
+
+      - name: Publish release distributions to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```

This might be a good place to give out a hardening recommendation:
```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1  # update `release/v1` to a commit SHA if you use Dependabot
```

---

No objection, although I'm not sure if there's precedent in the starter workflows for this. I'll look.

---

My understanding is that they usually demand commit SHA in all starter workflows which is what got us into this mess in the first place. So I figured this might be a good middle ground 🤷‍♂️

---

Yeah, looks like `gem-push.yml` has a commit SHA, and a comment suggesting a change to a symbolic ref for automatic changes. I'll do something similar here 🙂

---

Did a version of what `gem-push.yml` does, PTAL!

---

Personally, I prefer the rolling branch but I don't think the GH folks would agree to that.. Dependabot wouldn't update it, though (until converted to a tag or a commit hash).
I suppose your variant will work for most.